File: projects/PWM/src/common/protocols/src/slip.rs
const END: u8     = 0xC0;
const ESC: u8     = 0xDB;
const ESC_END: u8 = 0xDC;
const ESC_ESC: u8 = 0xDD;

#[derive(Debug)]
pub enum SlipErrorCode {
    BadEsc,
    BufferFull,
}

#[derive(Debug)]
pub struct SlipError {
    pub pos: usize,
    pub code: SlipErrorCode,
}

impl From<SlipErrorCode> for SlipError {
    fn from(err: SlipErrorCode) -> Self {
        Self {pos: 0, code: err}
    }
}

pub struct BasicBuffer<const CAPACITY: usize> {
    idx: usize,
    buf: [u8; CAPACITY],
}

pub struct Decoder<const CAPACITY: usize> {
    buf: BasicBuffer<CAPACITY>,
    is_escaping: bool
}

pub struct Encoder<const CAPACITY: usize> {
    buf: BasicBuffer<CAPACITY>,
}

////////////////////////////////////////////

impl<const CAPACITY: usize> BasicBuffer<CAPACITY> {
    pub fn new() -> Self {
        Self {idx: 0, buf: [0u8; CAPACITY] }
    }

    pub fn reset(&mut self) {
        self.idx = 0;
    }

    pub fn slice(&self) -> &[u8] {
        &self.buf[..self.idx]
    }

    pub fn put(&mut self, c: u8)-> Result<(), SlipErrorCode> {
        if self.idx >= CAPACITY {
            Err(SlipErrorCode::BufferFull)
        }

        else {
            self.buf[self.idx] = c;
            self.idx+=1;
            Ok(())
        }
    }
}

////////////////////////////////////////////

impl<const CAPACITY: usize> Decoder<CAPACITY> {
    pub fn new() -> Self {
        Self {
            buf: BasicBuffer::new(),
            is_escaping: false
        }
    }

    pub fn reset(&mut self) {
        self.is_escaping = false;
        self.buf.reset();
    }

    pub fn slice(&self) -> &[u8] {
        self.buf.slice()
    }

    pub fn feed(&mut self, input: &[u8]) -> Result<(usize, bool), SlipError> {
        let mut i = 0;

        while i < input.len() {
            let c = input[i]; // Consume 1 char from input buffer
            i += 1;           // Increment counter
            
            if self.is_escaping {
                self.is_escaping = false;
                match c {
                    ESC_END => {
                        match self.buf.put(END) {
                            Ok(_) => {},
                            Err(code) => {return Err(SlipError{ pos: i, code: code});}
                        }
                    }

                    ESC_ESC => {
                        match self.buf.put(ESC) {
                            Ok(_) => {},
                            Err(code) => {return Err(SlipError{ pos: i, code: code});}
                        }
                    }

                    _ => {return Err(SlipError{ pos: i, code: SlipErrorCode::BadEsc});}
                }
            }

            else {
                match c {
                    END => {
                        return Ok((i, true));
                    }

                    ESC => {
                        self.is_escaping = true;
                    }

                    // otherwise, put stuff in buffer
                    _ => {
                        match self.buf.put(c) {
                            Ok(_) => {},
                            Err(code) => {return Err(SlipError{ pos: i, code: code});}
                        }
                    }
                }
            }
        }

        // Input buffer processed, but no packet end yet detected
        Ok((i, false))
    }
}

impl<const CAPACITY: usize> Encoder<CAPACITY> {
    pub fn new() -> Self {
        Self {
            buf: BasicBuffer::new(),
        }
    }

    pub fn reset(&mut self) {
        self.buf.reset();
    }

    pub fn slice(&self) -> &[u8] {
        self.buf.slice()
    }

    pub fn feed(&mut self, input: &[u8]) -> Result<usize, SlipError>{
        let mut i = 0;
        while i < input.len() {
            let c = input[i];
            i += 1;

            match c {
                END => {
                    match self.buf.put(ESC) {
                        Ok(_) => {},
                        Err(code) => {return Err(SlipError{pos: i, code: code})}
                    }

                    match self.buf.put(ESC_END) {
                        Ok(_) => {},
                        Err(code) => {return Err(SlipError{pos: i, code: code})}
                    }
                },

                ESC => {
                    match self.buf.put(ESC) {
                        Ok(_) => {},
                        Err(code) => {return Err(SlipError{pos: i, code: code})}
                    }

                    match self.buf.put(ESC_ESC) {
                        Ok(_) => {},
                        Err(code) => {return Err(SlipError{pos: i, code: code})}
                    }
                },

                other => {
                    match self.buf.put(other) {
                        Ok(_)     => {}
                        Err(code) => {return Err(SlipError{pos: i, code: code})}
                    }
                }
            }
        }

        Ok(i)
    }

    pub fn finish(&mut self) -> Result<(), SlipErrorCode>{
        Ok(self.buf.put(END)?)
    }
}

File: projects/uart/examples/protocol/slip.py
"""
=================
Simple SLIP codec
=================

:Authors: - Florian Dupeyron <florian.dupeyron@mugcat.fr>
:Date: February 2023
"""

# Constants
END     = 0xC0
ESC     = 0xDB
ESC_END = 0xDC
ESC_ESC = 0xDD


# ┌────────────────────────────────────────┐
# │ Encode function                        │
# └────────────────────────────────────────┘

def encode(buffer: bytes, prepend_end: bool = False) -> bytes:
    """
    Encode a given buffer in the SLIP format
    :param buffer: Input buffer to convert
    :param prepend_end: Append END character at beginning of encoded buffer?

    :return: Encoded bytes
    """

    o = bytearray([] if not prepend_end else [END])

    i = 0
    while i < len(buffer):
        c = buffer[i]
        i += 1

        if c == END:
            o.append(ESC)
            o.append(ESC_END)
        elif c == ESC:
            o.append(ESC)
            o.append(ESC_ESC)
        else:
            o.append(c)

    # Append packet end
    o.append(END)
    
    return bytes(o)


# ┌────────────────────────────────────────┐
# │ Decoder class                          │
# └────────────────────────────────────────┘

class Decoder:
    """
    Stateful class to decode incoming SLIP data
    """

    def __init__(self):
        self.buf         = bytearray([])
        self.is_escaping = False

    def reset(self):
        self.buf         = bytearray([])
        self.is_escaping = False

    def feed(self, buffer: bytes) -> (int, bool):
        """
        Feeds decoder and decode a given SLIP buffer into raw data

        :param buffer: Input buffer to decode
        :return: The number of bytes consumed from input, and if a frame is available
        """

        i = 0
        while i < len(buffer):
            c  = buffer[i]
            i += 1

            if self.is_escaping:
                self.is_escaping = False

                if c == ESC_END:
                    self.buf.append(END)
                elif c == ESC_ESC:
                    self.buf.append(ESC)
                else:
                    raise ValueError(f"Unknown escape character code: {c}")
            else:
                if c == END:
                    return (i, True,)

                elif c == ESC:
                    self.is_escaping = True

                else:
                    self.buf.append(c)

        # Input buffer processed, but no packet end yet detected
        return (i, False)


# ┌────────────────────────────────────────┐
# │ Unitary tests                          │
# └────────────────────────────────────────┘

if __name__ == "__main__":
    in_buf  = bytes([0x01, 0x02, END, ESC, 0xFF])
    out_buf = encode(in_buf)

    print(in_buf.hex(":", 1))
    print(out_buf.hex(":", 1))

    decoder = Decoder()

    nbytes, this_is_the_end = decoder.feed(out_buf)
    assert(nbytes          == len(out_buf))
    assert(this_is_the_end == True       )

    print(decoder.buf.hex(":", 1))

    ###

    good = False
    try:
        decoder.reset()
        decoder.feed(bytes([0x00, ESC, 0x02]))
    except ValueError as exc:
        good = True
    finally:
        if not good:
            raise AssertionError("Exception not thrown")
